### PR TITLE
feat: response caching

### PR DESCRIPTION
**Pull request by anhnh12**
**What does this PR do?**
-   [x] Implement memory caching for API response
- [x] Currently apply for watchlist and ticker/compare

### CODE REVIEW

**Review by trkhoi — APPROVED** [review]

File: src/commands/defi/token/compare.ts [hunk 1]
```diff
@@ -105,18 +112,28 @@ const handler: CommandChoiceHandler = async (msgOrInteraction) => {
...
-      query: baseSymbol,
-      default_ticker: baseId,
-    }),
-    config.setGuildDefaultTicker({
-      guild_id: i.guildId ?? "",
-      query: targetSymbol,
-      default_ticker: targetId,
-    }),
-  ])
+  const { ok: ok1 } = await config.setGuildDefaultTicker({
```
Comment (tuanddd) [inline src/commands/defi/token/compare.ts:115]:
```suggestion
  const { ok: setDefaultBaseTickerOk } = await config.setGuildDefaultTicker({
```

File: src/commands/defi/token/compare.ts [hunk 2]
```diff
@@ -105,18 +112,28 @@ const handler: CommandChoiceHandler = async (msgOrInteraction) => {
...
+    query: baseSymbol,
+    default_ticker: baseId,
+  })
+  if (ok1) {
+    CacheManager.findAndRemove(
+      "ticker",
+      `ticker-default-${i.guildId}-${baseSymbol}`
+    )
+  }
+  const { ok: ok2 } = await config.setGuildDefaultTicker({
```
Comment (tuanddd) [inline src/commands/defi/token/compare.ts:126]:
```suggestion
  const { ok: setDefaultTargetTickerOk  } = await config.setGuildDefaultTicker({
```

File: src/commands/defi/token/compare_slash.ts [hunk 3]
```diff
@@ -106,18 +107,33 @@ const handler: CommandChoiceHandler = async (msgOrInteraction) => {
...
-      query: baseSymbol,
-      default_ticker: baseId,
-    }),
-    config.setGuildDefaultTicker({
-      guild_id: i.guildId ?? "",
-      query: targetSymbol,
-      default_ticker: targetId,
-    }),
-  ])
+  const { ok: ok1 } = await config.setGuildDefaultTicker({
```
Comment (tuanddd) [inline src/commands/defi/token/compare_slash.ts:110]: same

File: src/commands/defi/token/compare_slash.ts [hunk 4]
```diff
@@ -106,18 +107,33 @@ const handler: CommandChoiceHandler = async (msgOrInteraction) => {
...
+    query: baseSymbol,
+    default_ticker: baseId,
+  })
+  if (ok1) {
+    CacheManager.findAndRemove(
+      "ticker",
+      `ticker-default-${i.guildId}-${baseSymbol}`
+    )
+  }
+  const { ok: ok2 } = await config.setGuildDefaultTicker({
```
Comment (tuanddd) [inline src/commands/defi/token/compare_slash.ts:121]: same

File: src/utils/CacheManager.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    if (!val) {
+      val = await call()
+      cache.set(key, val, ttl)
+    }
+    return val as any
+  }
+
+  private find(pool: string, str: string) {
+    const cache = this.cachePools.get(pool)
+    return cache?.keys().filter((k) => k.includes(str))
```
Comment (tuanddd) [inline src/utils/CacheManager.ts:49]: chỗ `.includes` coi chừng về sau có khả năng tạo bug do nó match substring, an toàn hơn chút nhưng vẫn flexible thì nên dùng `startsWith`

**Review by tuanddd — CHANGES_REQUESTED** [review]

File: src/commands/defi/ticker_slash.ts [hunk 6]
```diff
@@ -47,11 +54,11 @@ async function renderHistoricalMarketChart({
...
   const currency = "usd"
-  const { ok, data } = await Defi.getHistoricalMarketData(
-    coinId,
-    currency,
-    days || 7
-  )
+  const { ok, data } = await CacheManager.get({
+    pool: "ticker",
+    key: `ticker-getHistoricalMarketData-${coinId}-${currency}-${days}`,
+    call: () => defi.getHistoricalMarketData(coinId, currency, days),
```
[the remark this answers is not in the document]
Comment (anhnh12) [reply src/commands/defi/ticker_slash.ts:60]: already had fallback value `days = 7`

**Review by tuanddd — APPROVED** [review]

**Pull request merged**